### PR TITLE
feat: support regex in ape.reverts()

### PR DESCRIPTION
**Pull request by helloibis**
### What I did
- Added regex support in `ape.reverts()`
- Cleared up `RevertsContextManager` tests and test docstrings

fixes: #873 
fixes: APE-311

### How I did it
Allow `expected_message` and `dev_message` in `RevertsContextManager` to accept `re.Pattern` objects

### How to verify it
```python
# Matches explictly "dev: foo" or "dev: bar"
with ape.reverts(re.compile(r"^dev: (foo|bar)$")):
    ...

# Matches explictly "dev: foo" or "dev: bar"
with ape.reverts(dev_message=re.compile(r"^dev: (foo|bar)$")):
    ...
```

### Checklist
- [x] All changes are completed
- [x] New test cases have been added
- [x] Documentation has been updated

### CODE REVIEW

File: src/ape/pytest/contextmanagers.py [hunk 1]
```diff
@@ -75,7 +80,11 @@ def _check_dev_message(self, exception: ContractLogicError):
         if offending_source is None or offending_source.line_start is None:
             raise AssertionError("Could not find line that caused revert.")
 
-        assertion_error_prefix = f"Expected dev revert message '{self.dev_message}'."
+        assertion_error_prefix = "Expected dev revert message '{}.'".format(
```
Comment (unparalleled-js) [inline src/ape/pytest/contextmanagers.py:83]: This is probably just a personal preference, but I find f-strs more readable than `.format()` strings.

```python
message = self.dev_message.pattern if isinstance(self.dev_message, re.Pattern) else self.dev_message
assertion_error_prefix = f"Expected dev revert message '{message}.'"
```

File: tests/functional/test_revert_context_manager.py [hunk 2]
```diff
@@ -43,32 +63,59 @@ def test_revert_msg_raises_partial(owner, reverts_contract_instance):
...
+    Test that ``AssertionError`` is raised if the contract dev message does not match the supplied
+    dev revert pattern.
     """
+    with pytest.raises(AssertionError):
+        with reverts(dev_message=re.compile(r"dev: [^ero]+")):
+            reverts_contract_instance.revertStrings(2, sender=owner)
+
 
+@geth_process_test
+def test_both(owner, reverts_contract_instance, geth_provider):
```
Comment (unparalleled-js) [inline tests/functional/test_revert_context_manager.py:115]:
```suggestion
def test_both_message_and_dev_str(owner, reverts_contract_instance, geth_provider):
```

or something a ta more

**Review by unparalleled-js — APPROVED** [review]
Looks good! A couple small suggestions but nothing blocking.

**Review by unparalleled-js — APPROVED** [review]

**Comment by unparalleled-js** [comment]
Just had this thought: If the user provides a normal message but the assertion fails, I wonder if we should go ahead and check if it is a regex string and re-try the assertion. Thoughts?

I am worried users may get confused when regex-strings are not working.

**Review by NotPeopling2day — APPROVED** [review]

**Comment by helloibis** [comment]
> Just had this thought: If the user provides a normal message but the assertion fails, I wonder if we should go ahead and check if it is a regex string and re-try the assertion. Thoughts?
> 
> I am worried users may get confused when regex-strings are not working.

@unparalleled-js The issue is we can't really check a regex without either having them pass in `re.Pattern` or using a different kwarg name (which is what Brownie does, e.g. `dev_msg_pattern` or similar)

If someone were to pass in a string that happened to contain regex characters like `?` or `.`, we can't really tell if they were trying to use regex or trying to actually use a string

For example `"test?"` could be passed in as a string by accident, but let's say the dev comment in the contract is `"test"`. If we treated that as regex then, it would actually pass the assertions even though it shouldn't. It seems like that could create confusion about what should or shouldn't be matched, since we're not using Brownie's method of using differently named kwargs

**Pull request merged**